freezer/freezer_url.go

```diff
@@ -139,26 +152,51 @@ func newFreezerSource(u *url.URL) (substrate.AsyncMessageSource, error) {
 			}
 		}
 		u1.RawQuery = newVals.Encode()
+		store, err = strawOpen(u1.String())
 
-		ss, err := strawOpen(u1.String())
-		if err != nil {
-			return nil, err
-		}
-		conf := AsyncMessageSourceConfig{
-			StreamStore: ss,
-			FreezerConfig: freezer.MessageSourceConfig{
-				Path:            u.Path,
-				PollPeriod:      10 * time.Second,
-				CompressionType: ct,
-			},
-		}
-		return sourcer(conf)
 	default:
 		return nil, fmt.Errorf("unsupported scheme : %s", u.Scheme)
 	}
+	if err != nil {
+		return nil, err
+	}
 
+	conf := AsyncMessageSourceConfig{
+		StreamStore: store,
+		FreezerConfig: freezer.MessageSourceConfig{
+			Path:            u.Path,
+			PollPeriod:      10 * time.Second,
+			CompressionType: ct,
+		},
+	}
+	source, err := sourcer(conf)
+	if err != nil {
+		return nil, err
+	}
+	return &closeStoreSource{
+		AsyncMessageSource: source,
+		store:              store,
+	}, nil
 }
 
 var sourcer = NewAsyncMessageSource
 
+type closeStoreSource struct {
+	substrate.AsyncMessageSource
+	store straw.StreamStore
+}
+
+func (sink closeStoreSource) Close() (err error) {
+	return closeAll([]io.Closer{sink.AsyncMessageSource, sink.store})
+}
+
 var strawOpen = straw.Open
+
+func closeAll(closers []io.Closer) (outErr error) {
+	for _, c := range closers {
+		if err := c.Close(); err != nil {
+			outErr = multierror.Append(outErr, err)
```

Comment:
`closeAll` would be about 1 line longer if it didn't use `multierror`. Personally, I don't think it's worth the extra dependency.

Also, the call sites would be cleaner if this was variadic.